### ISSUE TITLE
Some `:read` flow steps are erroring after update to Ash 2.11.3

### ISSUE THREAD

**Issue by sswrk** [ash-project/ash#652]
**Describe the bug**

One of my flows errors after changing versions from 2.11.2 to 2.11.3. The latest version (2.11.4) doesn't work either

**To Reproduce**
1. I have a resource (let's call it `MyApp.Resource1`) that has a `:create` action. This action is manual (`MyApp.Actions.CreateResource1`).
2. The manual action `MyApp.Actions.CreateResource1` runs a flow `MyApp.Flows.CreateResource1` and passes all changeset attributes and arguments to it.
3. The `MyApp.Flows.CreateResource1` flow's steps look like that:

```
steps do
  transaction :maybe_create_resource_1, MyApp.Resource1 do
    branch :maybe_create_resource_2, expr(<some expression>) do
      run_flow :resource_2, MyApp.Flows.MaybeCreateResource2ForResource1 do
        input <some input>
      end
    end
    
    ...
  end
end
```
The `run_flow :resource_2` step is being run
4. The `MyApp.Flows.MaybeCreateResource2ForResource1` flow's steps look like that:

```
steps do
  branch :resource_2_by_id, expr(<some expression>) do
    read :existing_resource_2_by_id, MyApp.Resource2, :read do
      not_found_error? false
      get? true
      
      input <some input>
      
      tenant arg(:tenant)
    end
  end
  ...
end
```

Once it gets to the `:existing_resource_2_by_id` step, it errors:

```
** (Ash.Error.Unknown) Context: resolving data on perform MyApp.Resource1.create > resolving data on Transaction :maybe_create_resource_1
     
     Unknown Error
     
     Context: resolving intermediate_data on fetch MyApp.Resource2.read
     * Context: resolving intermediate_data on fetch MyApp.Resource2.read
     
     ** (ArgumentError) `nil` is not a Spark DSL module.
     
       nil.spark_dsl_config()
       (spark 1.1.18) lib/spark/dsl/extension.ex:134: Spark.Dsl.Extension.dsl!/1
       (spark 1.1.18) lib/spark/dsl/extension.ex:206: Spark.Dsl.Extension.fetch_opt/4
       (spark 1.1.18) lib/spark/dsl/extension.ex:172: Spark.Dsl.Extension.get_opt/5
       (ash 2.11.4) lib/ash/query/query.ex:302: Ash.Query.new/3
       (ash 2.11.4) lib/ash/query/query.ex:2587: Ash.Query.to_query/1
       (ash 2.11.4) lib/ash/query/query.ex:1425: Ash.Query.put_context/3
       (ash 2.11.4) lib/ash/actions/read.ex:2273: Ash.Actions.Read.run_before_action/1
       (ash 2.11.4) lib/ash/actions/read.ex:383: anonymous fn/2 in Ash.Actions.Read.as_requests/5
       (ash 2.11.4) lib/ash/engine/request.ex:1129: Ash.Engine.Request.do_try_resolve_local/4
       (ash 2.11.4) lib/ash/engine/request.ex:998: Ash.Engine.Request.do_try_resolve/5
       (elixir 1.14.2) lib/enum.ex:4751: Enumerable.List.reduce/3
       (elixir 1.14.2) lib/enum.ex:2514: Enum.reduce_while/3
       (ash 2.11.4) lib/ash/engine/request.ex:1041: Ash.Engine.Request.try_resolve_dependencies_of/3
       (ash 2.11.4) lib/ash/engine/request.ex:460: Ash.Engine.Request.store_dependency/4
       (ash 2.11.4) lib/ash/engine/request.ex:404: Ash.Engine.Request.send_field/3
       (ash 2.11.4) lib/ash/engine/engine.ex:849: Ash.Engine.notify_local_request/4
       (ash 2.11.4) lib/ash/engine/engine.ex:466: Ash.Engine.run_iteration/1
       (ash 2.11.4) lib/ash/engine/engine.ex:307: Ash.Engine.run_to_completion/1
       (ash 2.11.4) lib/ash/engine/engine.ex:252: Ash.Engine.do_run/2
```

**Expected behavior**
This should not error.

**Runtime**
 - Elixir version 1.14.2
 - Erlang version 25.1.2
 - OS MacOS Ventura 13.4.1
 - Ash version 2.11.3

**Additional context**
It seems that this bug was introduced in the commit https://github.com/ash-project/ash/commit/93d4f1fd26306940dbe4cbeccfe54d597f34427f

I put some inspects here and there and found out that `data` in `lib/ash/actions/read.ex:380` is 
```
%{
  actor: <some actor>,
  async?: false,
  authorize?: true,
  verbose?: false
}
```

and `path` in `lib/ash/actions/read.ex:195` is

```
[[:resource_2, :existing_resource_2_by_id]]
```

**Comment by sswrk** [comment]
Thank you for a quick reply!

I wrote a test that uses `Ash.Test.Flow.Org` and `Ash.Test.Flow.User` resources instead of `MyApp.Resource1` and `MyApp.Resource2` and weirdly enough, it passed with no error.

The data layer I'm using is AshPostgres (version 1.3.30).

Also, I've discovered that this bug does not only happen when using `run_flow` – it occurs also in the following scenario:

1. There's a `MyApp.Resource3` with the `:read` action:
```
read :read do
    primary? true

    argument :build, :map

    prepare build(arg(:build))
end
```

The `:build` argument is there to allow filtering in flows. 
2. The step in the flow that uses this action looks like that:
```
read :existing_resorces_3, MyApp.Resource3, :read do
  input %{
    build: %{
       filter: [
         resource_2_id: path(result(:resource_2), [:id]),
         status: [in: [:a, :b]]
       ]
    }
  }
  
  tenant arg(:tenant)
end
```
And it errors like that:
```
** (Ash.Error.Unknown) Context: resolving data on perform MyApp.Resource1.create > resolving data on Transaction :maybe_create_resource_1
     
     Unknown Error
     
     Context: resolving intermediate_data on fetch MyApp.Resource3.read
     * Context: resolving intermediate_data on fetch MyApp.Resource3.read
     
     ** (ArgumentError) `nil` is not a Spark DSL module.
     
       nil.spark_dsl_config()
       (spark 1.1.18) lib/spark/dsl/extension.ex:134: Spark.Dsl.Extension.dsl!/1
       (spark 1.1.18) lib/spark/dsl/extension.ex:206: Spark.Dsl.Extension.fetch_opt/4
       (spark 1.1.18) lib/spark/dsl/extension.ex:172: Spark.Dsl.Extension.get_opt/5
       (ash 2.11.4) lib/ash/query/query.ex:302: Ash.Query.new/3
       (ash 2.11.4) lib/ash/query/query.ex:2587: Ash.Query.to_query/1
       (ash 2.11.4) lib/ash/query/query.ex:1425: Ash.Query.put_context/3
       (ash 2.11.4) lib/ash/actions/read.ex:2273: Ash.Actions.Read.run_before_action/1
       (ash 2.11.4) lib/ash/actions/read.ex:383: anonymous fn/2 in Ash.Actions.Read.as_requests/5
       (ash 2.11.4) lib/ash/engine/request.ex:1129: Ash.Engine.Request.do_try_resolve_local/4
       (ash 2.11.4) lib/ash/engine/request.ex:998: Ash.Engine.Request.do_try_resolve/5
       (elixir 1.14.2) lib/enum.ex:4751: Enumerable.List.reduce/3
       (elixir 1.14.2) lib/enum.ex:2514: Enum.reduce_while/3
       (ash 2.11.4) lib/ash/engine/request.ex:1041: Ash.Engine.Request.try_resolve_dependencies_of/3
       (ash 2.11.4) lib/ash/engine/request.ex:460: Ash.Engine.Request.store_dependency/4
       (ash 2.11.4) lib/ash/engine/request.ex:404: Ash.Engine.Request.send_field/3
       (ash 2.11.4) lib/ash/engine/engine.ex:849: Ash.Engine.notify_local_request/4
       (ash 2.11.4) lib/ash/engine/engine.ex:466: Ash.Engine.run_iteration/1
       (ash 2.11.4) lib/ash/engine/engine.ex:307: Ash.Engine.run_to_completion/1
       (ash 2.11.4) lib/ash/engine/engine.ex:252: Ash.Engine.do_run/2
```
I modified the flow step to not use the `:build` argument and created the `read_by_resource_2_id_and_allowed_statuses` action... and it worked. But what's weird is that there are many other actions (in other resources) that have the `:build` argument, and they are being used by flows with no errors 🤔 

What's even weirder is that the `:existing_resource_2_by_id` step from `MyApp.Flows.MaybeCreateResource2ForResource1` flow (the one that is run through the `run_flow` step) also uses an action with `:build` argument, BUT unlike in the case described above, modifying it to use a specialized action with no `:build` argument didn't result in fixing the error.

I'm not sure how to join the dots and proceed with this one.

**Comment by zachdaniel** [comment]
This is definitely strange. The only lead I currently have is that maybe this is returning `nil`? Could you inspect that line's result when you get your failure case? https://github.com/ash-project/ash/blob/main/lib/ash/flow/executor/ash_engine.ex#L943

**Comment by sswrk** [comment]
It seems that this line is never reached.

**Comment by marmor157** [comment]
@zachdaniel I've been able to recreate this inside the unit test, it's available in my fork https://github.com/marmor157/ash/tree/marmor157/652 

Here are the files in case you don't necessarily want to clone another git repo:

`issue652.ex`
```elixir
defmodule Ash.Test.ParentResource do
  use Ash.Resource,
    authorizers: [Ash.Policy.Authorizer]

  actions do
    create :create do
      primary? true
    end

    read :read do
      primary? true

      argument :build, :map

      prepare build(arg(:build))
    end

    update :disable do
      argument :status, :atom do
        constraints one_of: [:canceled, :removed]
        default :canceled
      end

      change set_attribute(:status, :removed) do
        where [attribute_equals(:status, :canceled)]
      end

      accept [:status]

      manual Ash.Test.Actions.CancelParentResource
    end

    update :system_update

    update :system_disable do
      validate one_of(:status, [:canceled, :removed])

      change set_attribute(:status, :canceled) do
        where [absent([:status])]
      end

      accept [:status]
    end
  end

  code_interface do
    define_for Ash.Test.Flow.Api
    define :cancel, action: :disable
  end

  attributes do
    uuid_primary_key :id

    attribute :status, :atom do
      allow_nil? false
    end

    create_timestamp :created_at do
      private? false
    end

    update_timestamp :updated_at do
      private? false
    end
  end

  relationships do
    has_many :child_resources, Ash.Test.ChildResource
  end

  policies do
    policy action(:create) do
      authorize_if always()
    end

    policy action(:read) do
      authorize_if always()
    end

    policy action(:system_update) do
      authorize_if always()
    end

    policy action(:disable) do
      authorize_if always()
    end

    policy action(:system_disable) do
      authorize_if always()
    end
  end
end

defmodule Ash.Test.ChildResource do
  use Ash.Resource,
    authorizers: [Ash.Policy.Authorizer]

  actions do
    create :create do
      primary? true

      argument :parent_resource, Ash.Test.ParentResource

      change manage_relationship(:parent_resource, type: :append)
    end

    update :system_cancel do
      change set_attribute(:status, :canceled)

      reject :all
    end

    read :read do
      primary? true
      argument :build, :map

      prepare build(arg(:build))
    end
  end

  policies do
    policy action(:read) do
      authorize_if always()
    end

    policy action(:system_cancel) do
      authorize_if always()
    end
  end

  attributes do
    uuid_primary_key :id

    attribute :status, :atom do
      default :active
      allow_nil? false
    end

    create_timestamp :created_at do
      private? false
    end

    update_timestamp :updated_at do
      private? false
    end
  end

  relationships do
    belongs_to :parent_resource, Ash.Test.ParentResource do
      allow_nil? false
    end
  end
end

defmodule Ash.Test.Flows.CancelParentResource do
  @moduledoc false

  use Ash.Flow

  flow do
    api Ash.Test.Flow.Api

    argument :record, Ash.Test.ParentResource do
      allow_nil? false
    end

    argument :status, :atom do
      allow_nil? false
    end

    returns :parent_resource
  end

  steps do
    custom :record_id, fn input, _ -> {:ok, input} end do
      input path(arg(:record), [:id])
    end

    branch :maybe_cancel_child_resource, expr(not is_nil(^result(:record_id))) do
      read :child_resources, Ash.Test.ChildResource, :read do
        not_found_error? false

        input %{
          build: %{
            filter: [parent_resource_id: path(arg(:record), [:id])]
          }
        }
      end

      map :cancel_child_resources, result(:child_resources) do
        update :cancel_child_resource, Ash.Test.ChildResource, :system_cancel do
          record element(:cancel_child_resources)

          input %{actor: arg(:actor)}
        end
      end
    end

    update :parent_resource, Ash.Test.ParentResource, :system_disable do
      record arg(:record)
      wait_for [result(:cancel_child_resources)]

      input %{
        status: arg(:status)
      }
    end
  end
end

defmodule Ash.Test.Actions.CancelParentResource do
  @moduledoc false

  use Ash.Resource.ManualUpdate

  alias Ash.Test.Flows.CancelParentResource

  @impl true
  def update(%Ash.Changeset{data: record, arguments: %{status: status}}, _opts, _) do
    %Ash.Flow.Result{
      result: result,
      errors: errors,
      valid?: valid?
    } = CancelParentResource.run(record, status)

    case {result, errors, valid?} do
      {resource, [], true} ->
        {:ok, resource}

      {_, errors, _} when is_list(errors) and length(errors) > 0 ->
        {:error, errors}

      {_, _, false} ->
        {:error, "Invalid"}
    end
  end
end
```

`issue652_test.ex`
```elixir
defmodule Ash.Test.Issue652 do
  @moduledoc false
  use ExUnit.Case, async: false

  alias Ash.Test.Flow.Api
  alias Ash.Test.ParentResource

  describe "issue652" do
    setup do
      parent_resource =
        ParentResource
        |> Ash.Changeset.for_create(:create, %{status: :active})
        |> Api.create!()

      %{
        parent_resource: parent_resource
      }
    end

    @tag :"652"
    test "cancel all child resources", %{parent_resource: parent_resource} do
      Ash.Test.ChildResource
      |> Ash.Changeset.for_create(:create, %{parent_resource: parent_resource})
      |> Api.create!(authorize?: false)

      Ash.Test.ChildResource
      |> Ash.Changeset.for_create(:create, %{parent_resource: parent_resource})
      |> Api.create!(authorize?: false)

      {:ok, %{status: :canceled}} =
        ParentResource.cancel(parent_resource)
    end
  end
end
```

And I modified `test/support/flow/registry.ex` to be
```elixir
defmodule Ash.Test.Flow.Registry do
  @moduledoc false
  use Ash.Registry

  entries do
    entry(Ash.Test.Flow.User)
    entry(Ash.Test.Flow.Org)
    entry Ash.Test.ParentResource
    entry Ash.Test.ChildResource
  end
end
```

**Comment by zachdaniel** [comment]
Could you PR the failing tests? I'll merge them and fix that way. That is the most convenient for me.

**Comment by marmor157** [comment]
Sure, I will clean it up a bit then because it's a bit messy now

**Comment by marmor157** [comment]
https://github.com/ash-project/ash/pull/657

**Comment by zachdaniel** [comment]
Thanks for the reproduction! Fixed in 3c42062861ecb862d179175fcab4efdf5075e0ca
